### PR TITLE
PR: Use `requests` instead of `urllib` for updates

### LINKED ISSUE #21398: Replace `urllib` with `requests` package when checking for updates
It is desirable to migrate from deprecated `urlopen` and `urlretrieve` from `urllib` to using the `requests` package.

### PR DESCRIPTION
**Pull request by mrclary**
## Description of Changes

* Replaced the use of `urllib.urlopen` and `urllib.urlrequest` with `requests.get` when checking and downloading updates.
* SSL certificates are now verified by default when checking for updates. This was already the default for downloading updates.
* Some error messaging was updated to reflect possible SSL certificate verification errors.

### Issue(s) Resolved

Fixes #21398

### CODE REVIEW

File: spyder/workers/updates.py
Comment (mrclary) [inline spyder/workers/updates.py:248]: @ccordoba12 @CAM-Gerlach , this line is causing a syntax error for only Linux and Python 3.8 CI tests. Win and macOS seem to be okay. Any thoughts?

File: spyder/workers/updates.py
Comment (dalthviz) [inline spyder/workers/updates.py:31]: Should we keep mentioning here that the error is related generated while checking for Spyder updates? So something like:
 ```suggestion
    'SSL certificate verification failed while checking for Spyder updates.<br><br>'
```

File: spyder/workers/updates.py
Comment (dalthviz) [inline spyder/workers/updates.py:35]: Similar to the comment above this could be something like:
```suggestion
    'Unable to connect to the internet while checking for Spyder updates. <br><br>'
```

**Review by dalthviz — COMMENTED** [review]
Thanks @mrclary ! Left some suggestions regarding the error text when doing the initial check for updates (keep mentioning that the error is related to Spyder checking for updates). I'm missing to give this a manual check but seems like it should work :+1:

File: spyder/workers/updates.py
Comment (ccordoba12) [inline spyder/workers/updates.py:36]:
```suggestion
SSL_ERROR_MSG = _(
    'SSL certificate verification failed while checking for Spyder updates.'
    '<br><br>Please contact your network administrator for assistance.'
)

CONNECT_ERROR_MSG = _(
    'Unable to connect to the internet while checking for Spyder updates.'
    '<br><br>Make sure the connection is working properly.'
```

- Constants need to be call in caps (the exception is `logger` because we decided to use it like that).
- Fix text width to fit in 79 columns.

File: spyder/workers/updates.py
Comment (ccordoba12) [inline spyder/workers/updates.py:143]:
```suggestion
            error_msg = SSL_ERROR_MSG
```

File: spyder/workers/updates.py
Comment (ccordoba12) [inline spyder/workers/updates.py:146]:
```suggestion
            error_msg = CONNECT_ERROR_MSG
```

File: spyder/workers/updates.py
Comment (ccordoba12) [inline spyder/workers/updates.py:281]:
```suggestion
            error_msg = SSL_ERROR_MSG
```

File: spyder/workers/updates.py
Comment (ccordoba12) [inline spyder/workers/updates.py:284]:
```suggestion
            error_msg = CONNECT_ERROR_MSG
```

File: spyder/workers/updates.py
Comment (ccordoba12) [inline spyder/workers/updates.py:17]:
```suggestion
from requests.adapters import ConnectionError, SSLError
```

**Review by ccordoba12 — COMMENTED** [review]
Some small suggestions for you @mrclary, the rest looks good to me.

Also, could you add a test for `WorkerDownloadInstaller`? That can be similar to the ones we have for `WorkerUpdates` and could also test the progress reporting through the `sig_download_progress` signal.

**Comment by mrclary** [comment]
> Some small suggestions for you @mrclary, the rest looks good to me.
> 
> Also, could you add a test for `WorkerDownloadInstaller`? That can be similar to the ones we have for `WorkerUpdates` and could also test the progress reporting through the `sig_download_progress` signal.

Actually, I think the only things to test in the `WorkerDownloadInstaller` are downloading and progress reporting. This class doesn't really do anything with respect to various circumstances (except Win/mac and Lite/Full, but I don't think we need to worry about that).

Do you have a recommendation on how to go about this? I was thinking:

- Run `UpdateInstallerDialog.start_installation` but somehow mock the download url to point to some small test file somewhere. Or...
- Instantiate `WorkerDownloadInstaller` and `UpdateInstallation`, connect `sig_download_progress` signal, and mock the download url to point to some small test file somewhere. Or...
- Either of the above, but somehow mock the `requests.Response.iter_content` results.

**Comment by dalthviz** [comment]
Checking this locally I noticed a couple of things:

## Status bar text is not being updated when dismissing update:

* Checking the update behavior from 5.4.4 to 5.5.0 (when an update is detected the status bar text changes):

detect_update


* Simulating the update availability using this PR installer (changing the string version in the installed files):

update_pr


## Status bar text is not being updated in case an error happens while checking updates (no internet connection):

Seems like this was happening also before so just sharing (could be worthy to fix here or maybe in a follow-up PR after this one gets merged).

update_no_internet

* Simulating the update availability using the PR installer:

## Status bar text is being wrongly updated in case case an error happens while downloading the installer (no internet connection):

Seems like this was happening also before so again just sharing (as the issue above, it could be worthy to fix here or maybe in a follow-up PR after this one gets merged).

download_no_internet

**Comment by mrclary** [comment]
@dalthviz, thanks for reporting these issues. I don't know that this PR introduced these since I haven't touched any of the status update code, only replacing `urllib` with `requests` (and some minor message box messages). Have you tested these on the latest 5.x commit?

If these issues are also present with the latest 5.x commit, then I recommend that we address them in a separate PR.

**Comment by mrclary** [comment]
@dalthviz, thanks for reporting. Yes, I believe these may have been occurring before this PR. I recommend that we fix those in a separate PR. I may be able to address them next week before 5.5.0 release.

**Comment by mrclary** [comment]
Per discussion with @ccordoba12, we will not include unit testing for the `WorkerDownloadInstaller`. Testing the download of a release artifact will artificially affect our download count on GitHub metrics. To avoid this, a more involved solution is required that would otherwise delay releasing 5.5.0. Additionally, because the end-of-life for the 5.x branch is imminent, efforts at more comprehensive testing are better deferred to the 6.x branch.

**Review by ccordoba12 — APPROVED** [review]
Thanks @mrclary!

File: spyder/workers/updates.py
Comment (CAM-Gerlach) [inline spyder/workers/updates.py:149]: We could add one more block here to catch `HTTPError`:

```python
except HTTPError as err:
    error_msg = HTTP_ERROR_MSG.format(page.status_code)
    ...
```

with a friendly message, e.g. :

```python
HTTP_ERROR_MSG = _(
    'HTTP error {status_code} when checking for updates.'
    '<br><br>Make sure your connection is working properly.'
)
```

File: spyder/workers/updates.py
Comment (CAM-Gerlach) [inline spyder/workers/updates.py:36]: This should say something more like "Unable to connect to the Spyder update service." since this could be something specific to that site or due to a firewall, proxy, etc. issue rather than just internet connectivity

**Comment by mrclary** [comment]
@CAM-Gerlach, I'm sorry, I should have requested a review from you as well. I'll submit a follow-up PR with your suggestions.

**Comment by CAM-Gerlach** [comment]
No worries, I was away on my trip to the Python core dev sprint until yesterday, so I wasn't around anyway.

**Pull request merged**
